source/lac/trilinos_epetra_vector.cc

```diff
@@ -0,0 +1,529 @@
+      MPI_Comm comm = epetra_comm->GetMpiComm();
+      int send_buffer_size(off_proc_elements.n_elements());
+      int n_procs(0);
+      MPI_Comm_size(comm, &n_procs);
+      std::vector<int> recv_count(n_procs);
+      MPI_Allgather(&send_buffer_size, 1, MPI_INT, &recv_count, 1,
+                    MPI_INT, comm);
+      std::vector<int> displacement(n_procs,0);
+      for (int i=1; i<n_procs; ++i)
+        displacement[i] = displacement[i-1] + recv_count[i-1];
+      const unsigned int recv_buffer_size = displacement[n_procs-1] +
+                                            recv_count[n_procs-1];
+
+      // Write the indices in the send buffer
+      std::vector<types::global_dof_index> send_index_buffer(send_buffer_size);
+      elem = off_proc_elements.begin();
+      for (int i=0; i<send_buffer_size; ++i)
+        {
+          send_index_buffer[i] = *elem;
+          ++elem;
+        }
+      // Send the index of the off-processor elements
+      std::vector<types::global_dof_index> recv_index_buffer(recv_buffer_size);
+      MPI_Allgatherv(&send_index_buffer[0], send_buffer_size, DEAL_II_DOF_INDEX_MPI_TYPE,
+                     &recv_index_buffer[0], &recv_count[0], &displacement[0],
+                     DEAL_II_DOF_INDEX_MPI_TYPE, comm);
+
+      // Write the values in the send buffer
+      std::vector<double> send_value_buffer(send_buffer_size);
+      elem = off_proc_elements.begin();
+      for (int i=0; i<send_buffer_size; ++i)
+        {
+          send_value_buffer[i] = V[*elem];
+          ++elem;
+        }
+      // Send the value of the off-processor elements
+      std::vector<double> recv_value_buffer(recv_buffer_size);
+      MPI_Allgatherv(&send_value_buffer[0], send_buffer_size, MPI_DOUBLE,
+                     &recv_value_buffer[0], &recv_count[0], &displacement[0], MPI_DOUBLE, comm);
+
+      // Write the data in the vector
+      for (unsigned int i=0; i<recv_buffer_size; ++i)
+        {
+          if (locally_owned_elements.is_element(recv_index_buffer[i]))
+            {
+              const TrilinosWrappers::types::int_type local_index =
+                vector->Map().LID(static_cast<TrilinosWrappers::types::int_type>(
+                                    recv_index_buffer[i]));
+              (*vector)[0][local_index] = recv_value_buffer[i];
+            }
+        }
+
+      return *this;
+    }
+
+
+
+    Vector &Vector::operator*= (const double factor)
+    {
+      AssertIsFinite(factor);
+      vector->Scale(factor);
+
+      return *this;
+    }
+
+
+
+    Vector &Vector::operator/= (const double factor)
+    {
+      AssertIsFinite(factor);
+      Assert(factor!=0., ExcZero());
+      *this *= 1./factor;
+
+      return *this;
+    }
+
+
+
+    Vector &Vector::operator+= (const VectorSpaceVector<double> &V)
+    {
+      // Check that casting will work.
+      Assert(dynamic_cast<const Vector *>(&V)!=NULL, ExcVectorTypeNotCompatible());
```

Comment:
The next line will always throw a ```std::bad_cast```. This ```Assert``` just performs an extra ```dynamic_cast``` so the error message is nicer.